File: java/parsers/src/main/java/jp/co/teruuu/parser/common/StringParser.java
package jp.co.teruuu.parser.common;

public class StringParser implements Parser<String> {
    public final String literal;
    public StringParser(String literal) {
        this.literal = literal;
    }

    @Override
    public ParseResult<String> parse(String input) {
        if(input.startsWith(literal)) {
            return new ParseResult.Success<>(literal, input.substring(literal.length()));
        }else {
            return new ParseResult.Failure<>("expect: " + literal, input);
        }
    }

}

File: scala/parsers/src/main/scala/jp/co/teruuu/parser/common/StringParser.scala
package jp.co.teruuu.parser.common

class StringParser(literal: String) extends Parser[String] {

  override def parse(input: String): ParseResult[String] = {
    input.startsWith(literal) match {
      case true => ParseOk(this.literal, input.substring(literal.length))
      case false => ParseNg("expect: " + this.literal, input)
    }
  }
}
